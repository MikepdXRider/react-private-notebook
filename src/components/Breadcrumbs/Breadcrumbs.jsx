import { Link } from 'react-router-dom';
import { useNoteFromPath } from '../../hooks/useNoteFromPath';

export default function Breadcrumbs({ className }) {
  const { currentNote, isEditing } = useNoteFromPath();

  const renderCurrentNote = (linkifyTitle) => {
    if (linkifyTitle) {
      return (
        <span>
          {' '}
          &raquo;{' '}
          <Link to={`/notes/${currentNote.id}`}>{currentNote.title}</Link>{' '}
          &raquo; Edit
        </span>
      );
    } else {
      return <span> &raquo; {currentNote.title}</span>;
    }
  };

  return (
    <nav className={className}>
      <Link to="/notes">Notebook</Link>
      {currentNote?.title && renderCurrentNote(isEditing)}
    </nav>
  );
}
