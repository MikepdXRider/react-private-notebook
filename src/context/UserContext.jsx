import { createContext, useContext, useMemo, useState } from 'react';
import { getUser } from '../services/users';

const UserContext = createContext();

const UserProvider = ({ children }) => {
  // Is this just for initializing? 
  // OR... When setUser is called, is currentUser called again? 
  const currentUser = getUser();
  console.log(currentUser);
  const [user, setUser] = useState(
    currentUser ? { id: currentUser.id, email: currentUser.email } : {}
    );
  console.log(user);

  const value = useMemo(() => ({ user, setUser }), [user]);

  return <UserContext.Provider value={value}>{children}</UserContext.Provider>;
};

const useUser = () => {
  const context = useContext(UserContext);

  if (context === undefined) {
    throw new Error('useUser must be used within a UserProvider');
  }

  return context;
};

export { UserContext, UserProvider, useUser };
