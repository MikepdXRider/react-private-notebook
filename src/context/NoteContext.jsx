import {
  createContext,
  useContext,
  useEffect,
  useReducer,
  useState,
} from 'react';
import {
  createNote,
  deleteNoteById,
  getNotes,
  updateNoteById,
} from '../services/notes';
import { parseNote, parseNotes } from '../utils/noteParser';
import { useUser } from './UserContext';

// payload is a note object:
// { title: String, content: String }
function notesReducer(notes, { type, payload }) {
  switch (type) {
    case 'create':
      return [payload, ...notes];
    case 'reset':
      return [...payload];
    case 'update':
      return notes.map((note) => (note.id === payload.id ? payload : note));
    case 'delete':
      return notes.filter((note) => note.id !== payload.id);
    default:
      throw Error(`Unknown action: ${type}`);
  }
}

export const NoteContext = createContext();

const NoteProvider = ({ children }) => {
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState('');
  const [notesStore, dispatch] = useReducer(notesReducer, []);
  const { user } = useUser();

  useEffect(() => {
    const fetchNotes = async () => {
      if (user.id) {
        const notes = await getNotes();
        dispatch({
          type: 'reset',
          payload: parseNotes(notes),
        });
        setLoading(false);
      }
    };
    fetchNotes();
  }, [user?.id]);

  const addNote = async ({ title, content }) => {
    setLoading(true);
    try {
      const [note] = await createNote({ userId: user.id, title, content });
      const payload = parseNote(note);
      dispatch({ type: 'create', payload });
      return payload;
    } catch (error) {
      setError(error.message);
    } finally {
      setLoading(false);
    }
  };

  const getNote = (id) => {
    return notesStore.find((note) => note.id === Number(id));
  };

  const updateNote = async ({ id, title, content }) => {
    setLoading(true);
    try {
      const [note] = await updateNoteById(id, { title, content });
      const payload = parseNote(note);
      dispatch({ type: 'update', payload });
      return parseNote(payload);
    } catch (error) {
      setError(error.message);
    } finally {
      setLoading(false);
    }
  };

  const deleteNote = async (id) => {
    setLoading(true);
    try {
      const [note] = await deleteNoteById(id);
      dispatch({ type: 'delete', payload: { id: note.id } });
      return parseNote(note);
    } catch (error) {
      setError(error.message);
    } finally {
      setLoading(false);
    }
  };

  return (
    <NoteContext.Provider
      value={{
        notes: { loading, error, data: notesStore },
        addNote,
        getNote,
        updateNote,
        deleteNote,
      }}
    >
      {children}
    </NoteContext.Provider>
  );
};

const useNotes = () => {
  const context = useContext(NoteContext);

  if (context === undefined) {
    throw new Error('useNotes must be used within a NoteProvider');
  }

  return context;
};

export { NoteProvider, useNotes };
