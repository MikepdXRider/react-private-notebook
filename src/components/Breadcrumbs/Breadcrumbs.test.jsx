import { render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { NoteProvider } from '../../context/NoteContext';
import { UserProvider } from '../../context/UserContext';
import Breadcrumbs from './Breadcrumbs';

jest.mock('../../context/UserContext');
jest.mock('../../context/NoteContext');

it('should display a link to the Notebook by default', () => {
  const { container } = render(
    <UserProvider>
      <NoteProvider>
        <MemoryRouter>
          <Breadcrumbs />
        </MemoryRouter>
      </NoteProvider>
    </UserProvider>
  );

  expect(container).toMatchSnapshot();
});

it('should display the current noteâ€™s title when viewing a note', () => {
  const { container } = render(
    <UserProvider mockUser={{ id: 1, email: 'test@example.com' }}>
      <NoteProvider mockNotes={[{ id: 1, title: 'Test Note!' }]}>
        <MemoryRouter initialEntries={['/notes/1']}>
          <Breadcrumbs />
        </MemoryRouter>
      </NoteProvider>
    </UserProvider>
  );

  expect(container).toMatchSnapshot();
});

it('should display "Edit" when editing a note', () => {
  const { container } = render(
    <UserProvider mockUser={{ id: 1, email: 'test@example.com' }}>
      <NoteProvider mockNotes={[{ id: 1, title: 'Test Note!' }]}>
        <MemoryRouter initialEntries={['/notes/1/edit']}>
          <Breadcrumbs />
        </MemoryRouter>
      </NoteProvider>
    </UserProvider>
  );

  expect(container).toMatchSnapshot();
});
