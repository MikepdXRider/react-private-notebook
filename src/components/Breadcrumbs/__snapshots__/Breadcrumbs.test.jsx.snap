// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should display "Edit" when editing a note 1`] = `
<div>
  <nav>
    <a
      href="/notes"
    >
      Notebook
    </a>
    <span>
       
      »
       
      <a
        href="/notes/1"
      >
        Test Note!
      </a>
       
      » Edit
    </span>
  </nav>
</div>
`;

exports[`should display a link to the Notebook by default 1`] = `
<div>
  <nav>
    <a
      href="/notes"
    >
      Notebook
    </a>
  </nav>
</div>
`;

exports[`should display the current note’s title when viewing a note 1`] = `
<div>
  <nav>
    <a
      href="/notes"
    >
      Notebook
    </a>
    <span>
       » 
      Test Note!
    </span>
  </nav>
</div>
`;
